### PR TITLE
Add cuda implementation for loss_mean_squared_per_channel_and_pixel

### PR DESCRIPTION
**Pull request by arrufat**
Thanks to @reunanen for helping out! :D

### CODE REVIEW

File: dlib/cuda/cuda_dlib.cu [hunk 1]
```diff
@@ -1833,6 +1833,21 @@ namespace dlib
...
 
+    // ----------------------------------------------------------------------------------------
+
+        __global__ void _cuda_compute_loss_mean_squared_per_channel_and_pixel(float* loss_out, float* g, const float* truth, const float* out_data, size_t n, const float scale)
+        {
+            float loss = 0;
+            for (auto i : grid_stride_range(0, n))
+            {
+                const float y = truth[i];
+                const float temp1 = y - out_data[i];
```
Comment (reunanen) [inline dlib/cuda/cuda_dlib.cu:1844]: Maybe just `temp` now? :)

**Comment by reunanen** [comment]
Did you eventually get some performance boost from using the CUDA version?

**Comment by arrufat** [comment]
Yes, there are some speed improvements. Here's the benchmark code:
```
#include <dlib/dnn.h>
#include <iostream>
#include <numeric>

namespace model
{
    using namespace dlib;
    using net_type =
        loss_mean_squared_per_channel_and_pixel<10, con<10, 1, 1, 1, 1, input_rgb_image>>;
}  // namespace model

auto main() -> int
try
{
    dlib::matrix<dlib::rgb_pixel> dummy(224, 224);
    dlib::assign_all_pixels(dummy, dlib::rgb_pixel(0, 0, 0));
    std::vector<dlib::matrix<dlib::rgb_pixel>> minibatch(1, dummy);
    std::array<dlib::matrix<float>, 10> label;
    for (auto& l : label)
    {
        l = dlib::matrix<float>(224, 224);
        dlib::assign_all_pixels(l, 0);
    }
    std::vector<std::array<dlib::matrix<float>, 10>> labels(1, label);
    dlib::resizable_tensor samples;
    model::net_type net;
    net.to_tensor(minibatch.begin(), minibatch.end(), samples);
    net(minibatch);
    std::cout << net << '\n';
    net.compute_loss(samples, labels.begin());
    std::array<float, 100> times;
    for (auto& time : times)
    {
        const auto t0 = std::chrono::high_resolution_clock::now();
        net.compute_loss(samples, labels.begin());
        const auto t1 = std::chrono::high_resolution_clock::now();
        const auto duration =
            std::chrono::duration_cast<std::chrono::duration<float, std::milli>>(t1 - t0).count();
        time = duration;
    }
    std::cout << "average time: " << std::fixed << std::setprecision(3)
              << std::accumulate(times.begin(), times.end(), 0.f, std::plus<float>()) /
                     times.size()
              << " ms\n";
}
catch (const std::exception& e)
{
    std::cout << e.what() << '\n';
}
```
- cpu version: 0.975 ms
- gpu version: 0.395 ms

Thank you again for your help :D

File: dlib/cuda/cuda_dlib.cu (hunk 1 above)
Comment (arrufat) [reply dlib/cuda/cuda_dlib.cu]: oops, yes! 😅

File: dlib/dnn/loss.h [hunk 2]
```diff
@@ -389,7 +389,8 @@ namespace dlib
     {
     public:
 
-        typedef weighted_label<unsigned long> training_label_type;
+        typedef dlib::weighted_label<unsigned long> weighted_label;
```
Comment (arrufat) [inline dlib/dnn/loss.h:392]: I added this typedef, because maybe users would expect this loss to behave as the `loss_multiclass_log_per_pixel_weighted_`.
Sorry for doing this in this PR...

Comment (davisking) [reply dlib/dnn/loss.h:392]: Na, seems good.  Update the _abstract.h docs too though.

File: dlib/cuda/cuda_dlib.cu [hunk 3]
```diff
@@ -1882,6 +1897,29 @@ namespace dlib
             loss = scale*floss;
         }
 
+        void compute_loss_mean_squared_per_channel_and_pixel::
+        do_work
+        (
```
Comment (davisking) [inline dlib/cuda/cuda_dlib.cu:1902]: Put the ( on the same line as `do_work`

File: dlib/dnn/loss.h [hunk 4]
```diff
@@ -3301,6 +3302,12 @@ namespace dlib
                 }
             }
 
+#ifdef DLIB_USE_CUDA
+            double loss;
+            cuda_compute(truth, output_tensor, grad, loss);
+            return loss;
+#else
+
```
Comment (davisking) [inline dlib/dnn/loss.h:3310]: Can you move the stuff in the `else` block into something analogous to the `cuda::compute_loss_mean_squared_per_channel_and_pixel` you added?  Then in the unit tests you can create both the cuda and cpu version, give them the same inputs, and check that both methods give identical results.

Comment (arrufat) [reply dlib/dnn/loss.h:3310]: Sure! Where should I put that? Inside this class or in `dlib/cuda/cpu_dlib.cpp` and then name it `cpu::compute_loss_mean_squared_per_channel_and_pixel`?

Comment (arrufat) [reply dlib/dnn/loss.h]: OK, I think I've done something sensible in the meantime (I think we have 13 hours of time difference).

I don't get identical results for cuda and cpu implementations. However, the difference between them is around 1e-7 or 1r-8. For this reason, I added a small tolerance margin in the tests of 1e-6.

Comment (davisking) [reply dlib/dnn/loss.h]: Sounds good

**Comment by davisking** [comment]
Looks good.  Thanks for another PR :)
